src/chains/filecoin/filecoin/src/blockchain.ts

```diff
@@ -63,56 +68,90 @@ export default class Blockchain extends Emittery.Typed<
-        blocks: [genesisBlock],
-        height: 0
-      })
-    );
-
     // Create the IPFS server
     this.ipfsServer = new IPFSServer(this.options.chain);
 
-    // Fire up the miner if necessary
-    if (this.options.miner.blockTime > 0) {
-      const intervalMine = () => {
-        this.mineTipset();
-      };
+    this.miningTimeout = null;
+
+    // We set these to null since they get initialized in
+    // an async callback below. We could ignore the TS error,
+    // but this is more technically correct (and check for not null later)
+    this.tipsetManager = null;
+    this.blockHeaderManager = null;
 
-      this.miningTimeout = setInterval(
-        intervalMine,
-        this.options.miner.blockTime * 1000
+    this.#database = new Database(options.database);
+    this.#database.once("ready").then(async () => {
+      this.blockHeaderManager = await BlockHeaderManager.initialize(
+        this.#database.blocks!
+      );
+      this.tipsetManager = await TipsetManager.initialize(
+        this.#database.tipsets!,
+        this.blockHeaderManager
       );
 
-      utils.unref(this.miningTimeout);
-    } else {
-      this.miningTimeout = null;
-    }
+      const recordedGenesisTipset = await this.tipsetManager.getTipsetWithBlocks(
+        0
+      );
+      if (recordedGenesisTipset === null) {
+        // Create genesis tipset
+        const genesisBlock = new BlockHeader({
+          ticket: new Ticket({
+            // Reference implementation https://git.io/Jt31s
+            vrfProof: this.rng.getBuffer(32)
+          }),
+          parents: [
+            // Both lotus and lotus-devnet always have the Filecoin genesis CID
+            // hardcoded here. Reference implementation: https://git.io/Jt3oK
+            new RootCID({
+              "/": "bafyreiaqpwbbyjo4a42saasj36kkrpv4tsherf2e7bvezkert2a7dhonoi"
+            })
+          ]
+        });
+
+        const genesisTipset = new Tipset({
+          blocks: [genesisBlock],
+          height: 0
+        });
+
+        this.tipsetManager.earliest = genesisTipset; // initialize earliest
+        await this.tipsetManager.putTipset(genesisTipset); // sets latest
+        await this.#database.db!.put(
+          "latest-tipset",
+          Quantity.from(0).toBuffer()
+        );
+      } else {
+        this.tipsetManager.earliest = recordedGenesisTipset; // initialize earliest
+        const data: Buffer = await this.#database.db!.get("latest-tipset");
+        const height = Quantity.from(data).toNumber();
+        const latestTipset = await this.tipsetManager.getTipsetWithBlocks(
+          height
+        );
+        this.tipsetManager.latest = latestTipset!; // initialize latest
+      }
 
-    setTimeout(async () => {
       await this.ipfsServer.start();
 
+      // Fire up the miner if necessary
+      if (this.options.miner.blockTime > 0) {
+        const intervalMine = () => {
+          this.mineTipset();
+        };
+
+        this.miningTimeout = setInterval(
```

Comment:
You should probably use a recursive `setTimeout` pattern here, it's what ganache has always done (and still does in v7). I think of this setting more like we are _targeting_ `blockTime` blocks, not to average 1 block every `blockTime`. in Ethereum-flavored ganache, if block `1` takes a long time to mine, block `2` won't be mined until `blockTime` seconds _after_ the previous block completed mining.

Reply:
I'm fine with doing it with `setTimeout`; makes more sense to be consistent with the Ethereum flavor. It also is likely more accurate how chains work anyway. This implementation was actually from  Tim and just got up for reviewal due to code moving.

Reply:
Since the mining logic was moved to a separate function as part of #842, can I implement the `setTimeout` logic change in that PR instead of this one?

Reply:
Deffered to #842: https://github.com/trufflesuite/ganache-core/pull/842#event-4466881006